Geoblocking ADR draft

mluk-sap opened the pull request:
<!-- Thank you for your contribution. Before you submit the pull request:
1. Follow contributing guidelines, templates, the recommended Git workflow, and any related documentation.
2. Read and submit the required Contributor Licence Agreements (https://github.com/kyma-project/community/blob/main/docs/contributing/02-contributing.md#agreements-and-licenses).
3. Test your changes and attach their results to the pull request.
4. Update the relevant documentation.
-->

**Description**

Changes proposed in this pull request:

- ...

**Pre-Merge Checklist**

- [ ] As a PR reviewer, verify code coverage and evaluate if it is acceptable.

**Related issues**
<!-- If you refer to a particular issue, provide its number. For example, `Resolves #123`, `Fixes #43`, or `See also #33`. -->

triffer commented on docs/assets/geoblocking-cr-controller.svg:
I think there will be no Kyma Geoblocking Operator, since it'S part of API Gateway module.

mluk-sap replied:
Fixed it by writing: `Geoblocking Operator (API Gateway Operator)`

triffer commented on docs/contributor/adr/0010-geoblocking-draft.md:
```diff
@@ -0,0 +1,387 @@
+- the Config Map containing custom IP range allow/block list is configured by the user, it becomes a contract
+- the Config Map containing IP range allow/block list downloaded from the SAP internal service is a Kyma internal thing and the implementation may change at any time, so no other module should use it
+
+Consequence: Config map capacity may be exceeded in future, which may require immediate attention.
```
A consequence might also be that we want to monitor the size of the config map and get notified once we are reaching the threshold.

mluk-sap replied:
Yes, I was also thinking about it. And it can be even done outside of ip-auth because everybody gets the same list of policies anyway. So we may be checking it even 'manually' in our local machines once per some time.

mluk-sap replied:
I added an additional sentence saying that someone (we) should monitor the size of the policy list.

barchw commented on docs/contributor/adr/0010-geoblocking-draft.md:
```diff
@@ -0,0 +1,387 @@
+- Ingress Gateway is created by default in the istio-system namespace
+- Users may influence Istio (via Gateway API) to create Ingress Gateway in different namespace
+- Authorization Policy references Ingress Gateway via selector, so they should be in the same namespace
+- Gateway CR is supposed to be a singleton (for now)
```
What Gateway CR? There are to many :D

mluk-sap replied:
Good point, I probably confused it with Geoblocking CR

barchw commented on docs/contributor/adr/0010-geoblocking-draft.md:
```diff
@@ -0,0 +1,387 @@
+
+Decision: Let's use a Config Map as a fallback and cache for IP range allow/block list. However, let's split it technically from the Config Map that contains IP ranges provided by the end-user, so:
+- the Config Map containing custom IP range allow/block list is configured by the user, it becomes a contract
+- the Config Map containing IP range allow/block list downloaded from the SAP internal service is a Kyma internal thing and the implementation may change at any time, so no other module should use it
```
There are more things to consider when going for ConfigMap as the backup/fallback:
- whether it should be accessible to be read (or should it be a secret)
- who should have access to this resource
- how to make sure that it is not being written to

mluk-sap replied:
As we discussed - I've checked this and there is no difference in RBAC between ConfigMap and Secret. This CM would be stored in the same namespace where ip-auth is deployed, so we benefit from 'namespace separation' and Pods (Service Accounts) in other namespaces won't see it 'by default'. But a power user with access to kyma-system namespace may reconfigure it. So it is rather not possible to 'additionally' protect it, as the RBAC rules are additive and there is no possibility to 'deny' access to that resource if something has bigger permissions already.

So this comes down to:
- ip-auth having a Role that has a read/write access to this particular single CM (with IP rules cache) in kyma-system in kyma-system namespace
- ip-auth having a Role that has a read access to CM in kyma-system to be able to read CM with IP rules provided by the end-user
- remaining rules - we don't touch, if something has access to CM in kyma-system namespace or all namespaces (like admin user) - then they are able to access it

I'll add few words about it

mluk-sap replied:
Added:
- a paragraph describing service accounts at the end of 'Namespaces' section
- additional sentence in 'IP range allow/block list fallback/caching' decision consequences

mluk-sap commented on docs/contributor/adr/0010-geoblocking-draft.md:
```diff
@@ -0,0 +1,387 @@
+
+The list of blocked IP ranges may be big (thousands of entries), so analyzing whether an IP address matches any IP range in the list is too time consuming.
+
+The list of IP ranges changes rarely, so it is better to build a data structure that supports quick comparision of IP addresses. The good candidate is a [radix tree](https://en.wikipedia.org/wiki/Radix_tree).
```
The problem is that hash table is not applicable here, because the list of policies contains IP **ranges** and we need to check whether a given single IP address is inside any of ranges. So we would need a hashtable with 2^32 IP addresses (for IPv4)...
And the functionality we need seems to be available OOTB here: https://pkg.go.dev/github.com/infobloxopen/go-trees/iptree#Tree.GetByIP

mluk-sap replied:
To underline - the ADR doesn't force any implementation here. It only suggests that the list of IP ranges may be huge and a linear search is not a good idea, because it has a complexity O(n). The radix tree was used as an example of something better than a 'raw' unsorted list.

Unfortunately, IP ranges may overlap, the IP range list is provided either by the user or by the SAP service and we don't control the data in any way. In particular, 'allowed' and 'blocked' ones may overlap as well (allowed take precedence). But AFAIR, it doesn't matter which particular rule was used to allow or block the traffic.

mluk-sap replied:
Okay, I adapted the description to underline that radix-tree is only a proposal, mentioned also about TSS and wrote that this may be decided during implementation phase:
```Because the list of IP ranges changes rarely (compared to the frequency of incoming connections) it is recommended to build some efficient data structure that supports quick comparision of IP addresses. The good candidate is a [radix tree](https://en.wikipedia.org/wiki/Radix_tree) or something like [TSS](https://citeseerx.ist.psu.edu/document?repid=rep1&type=pdf&doi=3aa1dd14e3d1c20d1f09b0ce0d4b4dd7b1190885). This can be decided during the implementation phase and would probably require some tests (benchmark, etc.).```

mluk-sap commented on docs/contributor/adr/0010-geoblocking-draft.md:
```diff
@@ -0,0 +1,387 @@
+- the Config Map containing custom IP range allow/block list is configured by the user, it becomes a contract
+- the Config Map containing IP range allow/block list downloaded from the SAP internal service is a Kyma internal thing and the implementation may change at any time, so no other module should use it
+
+Consequence: Config map capacity may be exceeded in future, which may require immediate attention.
```
Yeah, we were discussing multiple options here and we decided to not over-engineer it for now. Config map is super simple, available OOTB and may be used as backup and also as a cache. And we may revisit it later.

mluk-sap commented on docs/contributor/adr/0010-geoblocking-draft.md:
```diff
@@ -0,0 +1,387 @@
+
+#### Policy download optimization
+
+In order to reduce unnecessary updates of the list of blocked IP ranges the ip-auth service should use ETag mechanism, which is supported by the SAP internal service.
```
Yes, exactly, we are not allowed to refer to it

mluk-sap commented on docs/contributor/adr/0010-geoblocking-draft.md:
```diff
@@ -0,0 +1,387 @@
+In order to reduce unnecessary updates of the list of blocked IP ranges the ip-auth service should use ETag mechanism, which is supported by the SAP internal service.
+
+The local copy (stored in a Config Map) should be used to optimize the amount of connections to the SAP internal service:
+- if the local copy contains a newer version (lastUpdateTime) -> load it
```
> another pod already loaded the latest version
Yes, this is what I meant, I'll adapt it to be more descriptive :)

>a pod would always check first if the config map has a newer version than the one in memory
Yes, exactly. It is all about utilizing more the local copy instead of flooding SAP service with requests from all atomic pods. So the idea was that any pod checks for the update once per some time and either updates the whole list (if there is an update) or updates the lastUpdateTime to indicate that the check was performed and the list was still up to date.
And yes, it is theoretically possible that there is a new policy list in the local copy and even newer policy list in the geoblocking service, but in normal operation it is rather unlikely.
So I would say that it is an implementation detail - the code that looks at local copy and the one that looks at the SAP service may be completely separate, triggered by different timers, etc.

mluk-sap replied:
I adapted most of these points to be more precise, like:
```if the local copy contains a newer version of the policy list (lastUpdateTime attribute) -> it means that another pod has already downloaded it from the SAP internal service, so load the policy list from the local copy```

mluk-sap commented on docs/contributor/adr/0010-geoblocking-draft.md:
```diff
@@ -0,0 +1,387 @@
+
+This approach may cause issues if SAP internal service responds slowly or does not respond at all, because events would be consuming memory. System availability is a key factor here, so it is acceptable to drop events in order to ensure stability (prevent out of memory issues).
+
+Events should have some retry number, so the application can drop them if maximum number of retries is exceeded. Events should be also dropped if the queue is full. All such cases should be properly logged.
```
The requirement is to store this in persistent or in-memory storage for 'limited timeframe', so it doesn't give any particular numbers. Also - it will be highly dependent on the load caused by external users.

mluk-sap commented on docs/contributor/adr/0010-geoblocking-draft.md:
```diff
@@ -0,0 +1,387 @@
+- after making access decision the ip-auth should generate an event and store it in a memory queue
+- separate thread should grab events from the queue and send them to the SAP internal service.
+
+This approach may cause issues if SAP internal service responds slowly or does not respond at all, because events would be consuming memory. System availability is a key factor here, so it is acceptable to drop events in order to ensure stability (prevent out of memory issues).
```
I added an additional section about 'Graceful shutdown / probes' about this. I think the reasonable approach is to implement graceful shutdown, so the events are not lost during 'normal' operation. But we can't avoid losing some events is pod is crashed. It would be a price for 'simplicity'.

mluk-sap commented on docs/contributor/adr/0010-geoblocking-draft.md:
```diff
@@ -0,0 +1,387 @@
+
+| Resource                           | Who creates it                | Namespace                          |
+| ---------------------------------- | ----------------------------- | ---------------------------------- |
+| Geoblocking Operator               | Power user                    | kyma-system                        |
```
Fixed it by writing: `Geoblocking Operator (API Gateway Operator)`

mluk-sap commented on docs/contributor/adr/0010-geoblocking-draft.md:
```diff
@@ -0,0 +1,387 @@
+- if the update check has been recently performed (now - lastUpdateCheckTime < refreshInterval) -> don't check it again
+- if the update check hasn't been recently performed (now - lastUpdateCheckTime >= refreshInterval) -> check for an update
+- if there is a newer version -> download, apply it, update a local copy and set lastUpdateCheckTime and lastUpdateTime to the current time
+- if there is no newer version -> update lastUpdateCheckTime time (so other pods may skip the check)
```
fixed it: `if there is no newer version available in the SAP internal service`

pbochynski reviewed (commented):
Please use single file for drawio diagrams. The recommended approach is to use `{name}.drawio.svg` - it contain diagram sources and SVG representation in the single file.

mluk-sap commented:
> Please use single file for drawio diagrams. The recommended approach is to use `{name}.drawio.svg` - it contain diagram sources and SVG representation in the single file.

Thank you, it is much simpler :) I've just renamed the svg files and removed drawio ones.

mluk-sap commented on docs/contributor/adr/0010-geoblocking-draft.md:
```diff
@@ -0,0 +1,387 @@
+
+Decision: it doesn't make sense to expose so many detailed parameters, because the idea behind Geoblocking is to be compliant with regulations, which are usually very global (apply to the whole company / product / application).
+
+### IP range allow/block list fallback/caching
```
I added an explanation of 'fallback' and 'cache' in this section. Pls recheck :)

nataliasitko commented on docs/contributor/adr/0010-geoblocking.md:
```diff
@@ -0,0 +1,396 @@
+### Geoblocking CR and controller
+
+In order to ensure reliability and configurability a new Geoblocking Custom Resource and a new Geoblocking Controller is introduced. The controller is responsible for:
+- managing the ip-auth service deployment
```
```suggestion
- managing the ip-auth Service's Deployment
```

nataliasitko commented on docs/contributor/adr/0010-geoblocking.md:
```diff
@@ -0,0 +1,396 @@
+- performing configuration checks (like external traffic policy)
+- reporting geoblocking state
+
+There should be only one Geoblocking resource. The controller should set the additional Geoblocking CR in `Error` state, similarly to [APIGateway CR](https://github.com/kyma-project/api-gateway/blob/05ed57f3299f42c2565ed1b28b84dee808a5213b/controllers/operator/apigateway_controller.go#L100).
```
```suggestion
There should be only one GeoBlocking resource. The controller should set the additional Geoblocking CR in the `Error` state, similarly to [APIGateway CR](https://github.com/kyma-project/api-gateway/blob/05ed57f3299f42c2565ed1b28b84dee808a5213b/controllers/operator/apigateway_controller.go#L100).
```

nataliasitko commented on docs/contributor/adr/0010-geoblocking.md:
```diff
@@ -0,0 +1,396 @@
+
+| Resource                                    | Who creates it                 | Namespace                          |
+| ------------------------------------------- | ------------------------------ | ---------------------------------- |
+| Geoblocking Operator (API Gateway Operator) | Power user / Lifecycle Manager | kyma-system                        |
+| Geoblocking CR                              | Power user                     | kyma-system                        |
+| ip-auth Deployment                          | Geoblocking Operator           | kyma-system                        |
+| ip-auth Service                             | Geoblocking Operator           | kyma-system                        |
+| ip-auth Secret                              | Power user                     | kyma-system                        |
+| IP ranges Config Map with input             | ip-auth or External customer   | kyma-system                        |
+| IP ranges Config Map with cache             | ip-auth or External customer   | kyma-system                        |
+| Authorization Policy                        | Geoblocking Operator           | istio-system                       |
```
```suggestion
| GeoBlocking Operator (API Gateway Operator) | Power user / Lifecycle Manager | `kyma-system`                        |
| GeoBlocking CR                              | Power user                     | `kyma-system`                        |
| ip-auth Deployment                          | GeoBlocking Operator           | `kyma-system`                        |
| ip-auth Service                             | GeoBlocking Operator           | `kyma-system`                        |
| ip-auth Secret                              | Power user                     | `kyma-system`                        |
| IP ranges ConfigMap with input             | ip-auth or external customer   | `kyma-system`                        |
| IP ranges ConfigMap with cache             | ip-auth or external customer   | `kyma-system`                        |
| AuthorizationPolicy                        | GeoBlocking Operator           | `istio-system`                       |
```

nataliasitko commented on docs/contributor/adr/0010-geoblocking.md:
```diff
@@ -0,0 +1,396 @@
+
+#### Policy download optimization
+
+In order to reduce unnecessary updates of the list of blocked IP ranges the ip-auth service should use ETag mechanism, which is supported by the SAP internal service.
```
```suggestion
In order to reduce unnecessary updates to the list of blocked IP ranges, the ip-auth Service should use the ETag mechanism, which is supported by the SAP internal service.
```

nataliasitko commented on docs/contributor/adr/0010-geoblocking.md:
```diff
@@ -0,0 +1,396 @@
+
+Geoblocking mechanism touches multiple resources in multiple areas:
+- Istio configuration
+- Ingress gateway
```
```suggestion
- Ingress Gateway
```

nataliasitko commented on docs/contributor/adr/0010-geoblocking.md:
```diff
@@ -0,0 +1,396 @@
+- Ingress gateway
+- Gateways in general
+- Authorization Policy
+- Authorizer workload (service, deployment)
```
```suggestion
- Authorizer workload (Service, Deployment)
```

nataliasitko commented on docs/contributor/adr/0010-geoblocking.md:
```diff
@@ -0,0 +1,396 @@
+- Gateways in general
+- Authorization Policy
+- Authorizer workload (service, deployment)
+- Configuration of SAP internal service
```
```suggestion
- Configuration of the SAP internal service
```

nataliasitko commented on docs/contributor/adr/0010-geoblocking.md:
```diff
@@ -0,0 +1,396 @@
+
+## Context
+
+Geoblocking is a feature allowing to block incoming traffic coming from certain IP addresses, that are exclusive to certain countries or regions.
+Can be utilized against anonymous users and system to system network communication, where both are identified only by its source IP address.
+Many companies are using geoblocking to protect their services from unwanted traffic, such as DDoS attacks, or to comply with legal regulations.
+Therefore, creating this kind of feature is a convenience for the user, as it extracts the need to implement it on its own. 
```
We (try to) use camel case for Kubernetes resources' kinds in our docs like Deployments, Pods, Services, etc. So, I also used "GeoBlocking" when we talk about the CR or the name of the controller/operator but left "geoblocking" when it's about the feature.

mluk-sap commented on docs/contributor/adr/0010-geoblocking-draft.md:
```diff
@@ -0,0 +1,387 @@
+- service:
+  - selectors
+  - ports
+
```
Okay, added a short section 'Deployment stability' that mentions this among other features like autoscaling, probes, rolling updates, etc.

mluk-sap commented on docs/contributor/adr/0010-geoblocking.md:
```diff
@@ -0,0 +1,396 @@
+
+### IP range allow/block list fallback/caching
+
+The list of IP ranges is processed by each ip-auth container and stored in its memory and regularly refreshed using geoblocking service. If geoblocking service doesn't work then the ip-auth containers may work as before, the list is not refreshed.
```
GeoBlocking Service replaced with the SAP internal service

mluk-sap commented on docs/contributor/adr/0010-geoblocking.md:
```diff
@@ -0,0 +1,396 @@
+
+The list of IP ranges is processed by each ip-auth container and stored in its memory and regularly refreshed using geoblocking service. If geoblocking service doesn't work then the ip-auth containers may work as before, the list is not refreshed.
+
+However, the problem is when the ip-auth container needs to be restarted when geoblocking service is not available, so it can't download the list of IP ranges. This may be the case during an upgrade of ip-auth or in the event of runtime problems (for example ouf of memory). It is not acceptable to allow the traffic in such case. This means that the list of IP ranges should be stored somehow to be ...
```
GeoBlocking Service replaced with the SAP internal service

mluk-sap commented on docs/contributor/adr/0010-geoblocking.md:
```diff
@@ -0,0 +1,396 @@
+
+Decision: Don't implement additional mechanisms and follow the 'standard' approach, so the controller reports issues in the CR only from its own layer, while the ip-auth service reports issues in its log or via liveness/readiness probes (like every other workload).
+
+Consequence: Geoblocking CR won't be responsible for presenting Geoblocking 'health'. Support or developers must be aware that they should also inspect ip-auth logs in case of issues.
```
We discussed it, should be geoblocking health
